autorest/azure/example/main.go

```diff
@@ -17,83 +18,128 @@ import (
+
 	flag.Parse()
 
-	log.Println("Using these settings:")
-	log.Println("* certificatePath:", certificatePath)
-	log.Println("* applicationID:", applicationID)
-	log.Println("* tenantID:", tenantID)
-	log.Println("* subscriptionID:", subscriptionID)
+	log.Printf("mode(%s) certPath(%s) appID(%s) tenantID(%s), subID(%s)\n",
+		mode, certificatePath, applicationID, tenantID, subscriptionID)
 
-	if strings.Trim(certificatePath, " ") == "" ||
-		strings.Trim(applicationID, " ") == "" ||
-		strings.Trim(tenantID, " ") == "" ||
+	if strings.Trim(tenantID, " ") == "" ||
 		strings.Trim(subscriptionID, " ") == "" {
-		log.Fatalln("Bad usage. Please specify all four parameters")
+		log.Fatalln("Bad usage. Please specify applicationID, tenantID, subscriptionID")
+	}
+
+	if mode != "certificate" && mode != "cached" && mode != "device" {
+		log.Fatalln("Bad usage. Mode must be one of 'certificate', 'cached' or 'device'.")
+	}
+
+	if mode == "device" && applicationID == "" {
+		applicationID = xplatClientID
+	}
+
+	if mode == "certificate" && strings.Trim(certificatePath, " ") == "" {
+		log.Fatalln("Bad usage. Mode 'certificate' requires the 'certificatePath' argument.")
 	}
 }
 
-func main() {
-	log.Println("loading certificate... ")
+func getSptFromCachedToken(clientID, tenantID, resource string) (*azure.ServicePrincipalToken, error) {
+	token, err := azure.LoadToken(tokenCachePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load token from cache: %v", err)
+	}
+
+	spt := azure.NewServicePrincipalTokenFromManualToken(
+		clientID,
+		tenantID,
+		resource,
+		*token)
+
+	return spt, nil
+}
+
+func decodePkcs12(pkcs []byte, password string) (*x509.Certificate, *rsa.PrivateKey, error) {
+	privateKey, certificate, err := pkcs12.Decode(pkcs, password)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rsaPrivateKey, isRsaKey := privateKey.(*rsa.PrivateKey)
+	if !isRsaKey {
+		return nil, nil, fmt.Errorf("PKCS#12 certificate must contain an RSA private key")
+	}
+
+	return certificate, rsaPrivateKey, nil
+}
+
+func getSptFromCertificate(clientID, tenantID, resource, certicatePath string) (*azure.ServicePrincipalToken, error) {
 	certData, err := ioutil.ReadFile(certificatePath)
 	if err != nil {
-		log.Fatalln("failed", err)
+		return nil, fmt.Errorf("failed to read the certificate file (%s): %v", certificatePath, err)
 	}
 
 	certificate, rsaPrivateKey, err := decodePkcs12(certData, "")
 	if err != nil {
-		log.Fatalln("failed", err)
+		return nil, fmt.Errorf("failed to decode pkcs12 certificate while creating spt")
 	}
 
-	log.Println("retrieve oauth token... ")
-	spt, err := azure.NewServicePrincipalTokenFromCertificate(
-		applicationID,
+	spt := azure.NewServicePrincipalTokenFromCertificate(
+		clientID,
 		certificate,
 		rsaPrivateKey,
 		tenantID,
 		azure.AzureResourceManagerScope)
+
+	return spt, nil
+}
+
+func getSptFromDeviceFlow(clientID, tenantID, resource string) (*azure.ServicePrincipalToken, error) {
```

Comment:
You get back an SPT. That's how you get autorefreshing, etc. Also, SPT implements the interface needed to be assigned to `autorest.Client`'s `Authorizer` field